Handle DMI busy in sba write.

timsifive opened the pull request:
If we encounter DMI busy on the NOP after a read, we'll never get the
value out because DMI busy is sticky. The read must be retried, but we
don't know whether it was ever issued. Since the read has side effects
(incrementing of the address) this retry must be handled at a higher
layer. So now dmi_op_timeout can be told to retry or not, and if retry
is disabled it'll return an error when busy.

Also actually properly do the retry in dmi_op_timeout(). Previously the
code would not reissue the command and end up returning a garbage value.

This is a cleaner alternative to #434

timsifive committed:
Handle DMI busy in sba write.

If we encounter DMI busy on the NOP after a read, we'll never get the
value out because DMI busy is sticky. The read must be retried, but we
don't know whether it was ever issued. Since the read has side effects
(incrementing of the address) this retry must be handled at a higher
layer. So now dmi_op_timeout can be told to retry or not, and if retry
is disabled it'll return an error when busy.

Also actually properly do the retry in dmi_op_timeout(). Previously the
code would not reissue the command and end up returning a garbage value.

Change-Id: I3b52ebd51ebbbedd6e425676ac861b57fbe711b1

timsifive committed:
Fix whitespace.

Change-Id: Icb76d964e681b22346368d224d1930c9342343f3

timsifive committed:
Handle a few more DMI busy cases.

Change-Id: I8503a44e4bf935c0ebfff0d598fe4c322fda702a

GregSavin reviewed (dismissed):
So, calling dmi_op with retry=true is generally appropriate if and only if repeating the particular DMI operation being invoked is known not to have stateful side effects on the target?

timsifive commented:
> So, calling dmi_op with retry=true is generally appropriate if and only if repeating the particular DMI operation being invoked is known not to have stateful side effects on the target?

Yes, that's exactly right. Should I add a comment somewhere to make that more clear?

SiFiveGregS commented:
It might be worth including that in the comments (for new maintainers etc)

timsifive committed:
Explain when to use dmi_op_timeout(retry).

Change-Id: I1a5c6d76ac41a84472a8f79faecb2f48105191ff

GregSavin commented on src/target/riscv/riscv-013.c:
```diff
@@ -3266,9 +3274,14 @@ static int write_memory_progbuf(struct target *target, target_addr_t address,
-		if (dmi_op(target, &abstractcs, &dmi_busy_encountered, DMI_OP_READ,
-					DMI_ABSTRACTCS, 0, false, true) != ERROR_OK)
+		result = dmi_op(target, &abstractcs, DMI_OP_READ, DMI_ABSTRACTCS,
+				0, false, true, false);
```
Curious why retry is specified as false in this particular call, but then if the result comes back as busy, then a second read operation is made immediately, but with retry=true.  (Why not specify retry=true in the first call?  Is the first call variant more efficient, and usually expected not to have a busy result?)

JanMatCodasip commented:
> If we encounter DMI busy on the NOP after a read, we'll never get the
value out because DMI busy is sticky. The read must be retried, but we
don't know whether it was ever issued.

I am sorry but I believe this point is incorrect. If we issue a "DMI read" and get dmi.op==0 as a response in that very same DMI scan, we know that the new DMI read must have been started, because dmi.op==0 says all previous transactions had been finished by the time our new read request came. In short, we can be sure that our DMI read got really started.

I have elaborated more on this under #434 

So in the end, there should not be any need to expose the "BUSY" state and retry mechanisms to the upper layers above `dmi_op()/dmi_op_timeout()`. All the DTM retry logic can stay in `dmi_op_timeout()` and need not "pollute" the upper layers.

If there is a case when we issue a "DMI read", get back dmi.op==0 in the same DMI scan but the target does not carry out the DMI read, than in my opinion, this is a bug of the target and violation of the spec.

timsifive committed:
dmi_reset does not affect the current transaction.

That means the retry scheme we had been using works fine. This does
contain some minor tweaks, and now we pass my tests which hammer the DMI
busy case harder.

Change-Id: I13eee384dbba82bc5a5b1d387c75c547afe557b5

timsifive committed:
Remove unnecessary changes to make the PR readable

Change-Id: I87079876e6965563cf590e3936b3595aeab8715d

timsifive committed:
Move idle to end of line...

... because we go through run-test/idle after the scan.

Change-Id: I21a8cff22471f0b895d8cd8d25373dced9bf1ca9

GregSavin commented on src/target/riscv/riscv-013.c:
```diff
@@ -582,65 +584,73 @@ static int dmi_op_timeout(struct target *target, uint32_t *data_in,
-				} else if (status == DMI_STATUS_SUCCESS) {
-					return ERROR_OK;
+			if (status == DMI_STATUS_FAILED) {
+				if (data_in) {
```
Is the body of this "if" statement reachable?  I see the assignment to status in line 616, and if the value DMI_STATUS_FAILED gets assigned to status at line 616, then the "else" clause at line 624 will be executed, including the return statement on line 627, in which case the status variable seems guaranteed to be unequal to DMI_STATUS_FAILED at line 632.

timsifive replied:
You're right. Looks like some code got duplicated there. Thanks!

timsifive committed:
Remove unused code.

Change-Id: I07a7cdd2d64ca40a4fe181111a34cf55ff1928d1
